### DISCUSSION TITLE
Can performing Allreduce and computations in parallel reduce training speed?

### DISCUSSION THREAD

**Discussion by ihchoi12** [horovod/horovod#3577] [Q&A]
Hello team!

I'm analyzing Nsight Systems profile of my training benchmark.
I'm using two AWS p3.x2large nodes (1 GPU per node).

When I run the training using default configuration (`horovodrun -np 2 -H localhost:1,172.31.20.228:1 -p 12345 --verbose nsys profile -t cuda,osrt,nvtx,cudnn,cublas -f true -o ./baseline -w true  python /root/byteps/bytescheduler/examples/pytorch_horovod_benchmark.py --num-iters 2 --model resnet50 --num-classes 10000`), I see that calculating gradient takes about 160ms with lots of idle time in between the computation kernels:
<img width="1012" alt="image" src="https://user-images.githubusercontent.com/37887404/173276058-d5eebeb9-a57a-475d-b524-39c98de1a405.png">

Observing the idle times when the computations are performed with NCCL Allreduce operations, I tried to separate the NCCL Allreduce by configuring horovod cycle time and tensor fusion threshold (`horovodrun --cycle-time-ms 100 --fusion-threshold-mb 128 -np 2 -H localhost:1,172.31.20.228:1 -p 12345 --verbose nsys profile -t cuda,osrt,nvtx,cudnn,cublas -f true -o ./baseline -w true  python /root/byteps/bytescheduler/examples/pytorch_horovod_benchmark.py --num-iters 2 --model resnet50 --num-classes 10000`), and the result is like this:
<img width="1027" alt="image" src="https://user-images.githubusercontent.com/37887404/173276353-8a37c8c5-67f1-4542-9754-58909b3ff7b9.png">

Interestingly, I observe that it takes just 60ms for calculating gradients with removing most of the idle times.

So, I wonder if it means that launching the NCCL allreduce kernels in between the computation kernels makes the overall training speed lower, and how we can solve this issue to perform the communication and computation in parallel without overheads. 

Any comments will be much appreciated. Thanks!

**Comment by maxhgerlach** [comment]
Hi @ihchoi12, 
overlapping backward pass computation and gradient communication (of layers computed earlier) usually saves time in aggregate, but of course your mileage can vary: This could be different with your model, hardware and networking. Have you experimented with Horovod's grouped allreduce feature already? You could try and force all allreduces to be done in a single group -- that way all gradients are computed before the reduction begins, but you do not need to introduce extra latencies coming with long Horovod cycle times. In my experiments this noticeably slowed down training steps, though.
